nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/encrypt/KeyedCipherPropertyEncryptor.java

```diff
@@ -0,0 +1,160 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.encrypt;
+
+import org.apache.nifi.security.kms.CryptoUtils;
+import org.apache.nifi.security.util.EncryptionMethod;
+import org.apache.nifi.security.util.crypto.KeyedCipherProvider;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.SecretKey;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Property Encryptor implementation using Keyed Cipher Provider
+ */
+class KeyedCipherPropertyEncryptor extends CipherPropertyEncryptor {
+    private static final int INITIALIZATION_VECTOR_LENGTH = 16;
+
+    private static final int ARRAY_START = 0;
+
+    private static final boolean ENCRYPT = true;
+
+    private static final boolean DECRYPT = false;
+
+    private final SecureRandom secureRandom;
+
+    private final KeyedCipherProvider cipherProvider;
+
+    private final EncryptionMethod encryptionMethod;
+
+    private final SecretKey secretKey;
+
+    protected KeyedCipherPropertyEncryptor(final KeyedCipherProvider cipherProvider,
+                                           final EncryptionMethod encryptionMethod,
+                                           final SecretKey secretKey) {
+        this.cipherProvider = cipherProvider;
+        this.encryptionMethod = encryptionMethod;
+        this.secretKey = secretKey;
+        this.secureRandom = new SecureRandom();
+    }
+
+    /**
+     * Encrypt binary and return enciphered binary prefixed with initialization vector
+     *
+     * @param binary Binary to be encrypted
+     * @return Enciphered binary prefixed with initialization vector
+     */
+    @Override
+    protected byte[] encrypt(final byte[] binary) {
+        final byte[] initializationVector = createInitializationVector();
+        final Cipher cipher = getCipher(initializationVector, ENCRYPT);
+        try {
+            final byte[] encrypted = cipher.doFinal(binary);
+            return CryptoUtils.concatByteArrays(initializationVector, encrypted);
+        } catch (final BadPaddingException | IllegalBlockSizeException e) {
+            final String message = String.format("Encryption Failed with Algorithm [%s]", cipher.getAlgorithm());
+            throw new EncryptionException(message, e);
+        }
```

Comment:
Hi,
This code snippet also can be found PasswordBasedCipherPropertyEncryptor. Do you think it'd worth to extract this method into their common PropertyEncryptor interface as a default one?

Reply:
This method implementation is slightly different from `PasswordBasedCipherPropertyEncryptor` in that the `Keyed` implementation uses an initialization vector whereas the `PasswordBased` implementation uses a salt.  It is possible to name the reading of the parameter binary information more generically, which would allow removing this duplicated block.